### PR TITLE
dartsim 6.3.0 (new formula)

### PR DESCRIPTION
**Pull request by jslee02**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

DART provides data structures and algorithms for kinematic and dynamic applications in robotics and computer animation.

We use `dartsim` as the package name instead of `dart` to avoid name conflicts with [Dart Language](https://www.dartlang.org/).

Repo: https://github.com/dartsim/dart

Depends on #23902 and #22726.

### CODE REVIEW

**Comment by woodruffw** [comment]
@BrewTestBot test this please

**Comment by fxcoudert** [comment]
This needs `urdfdom`, which in turn needs `urdfdom_headers`. Given those aren't used outside of this formula, and they seem to have a relatively small user base in themselves, I'm wondering if we should just vendor them in the `dartsim` formula.

**Comment by jslee02** [comment]
[URDF](http://wiki.ros.org/urdf) is one of the largely used robot description format in the robotics field, which is maintained by [Open Source Robotics Foundation](https://www.osrfoundation.org/). I do believe many robotics libraries/software are using this library to parser URDF such as [Gazebo](http://gazebosim.org/), [OpenRAVE](http://openrave.org/), [Bullet](http://bulletphysics.org/wordpress/), and so on.

**Comment by scpeters** [comment]
I work at Open Source Robotics Foundation, and I help maintain urdfdom, urdfdom_headers, and console_bridge. These are core packages of Robot Operating System (ROS), which is very popular in robotics. The [ros_comm](https://github.com/ros/ros_comm) repository has lots of stars, and it depends on urdfdom, urdfdom_headers, and console_bridge.

I understand your point about vendoring if the formulae aren't very common, but I hope you will accept those other formulae into homebrew-core.

**Comment by j-rivero** [comment]
> URDF is one of the largely used robot description format in the robotics field, which is maintained by Open Source Robotics Foundation. I do believe many robotics libraries/software are using this library to parser URDF such as Gazebo, OpenRAVE, Bullet, and so on.

Correct. And I'm afraid that we have suffered from people embedding random copies of urdfdom (see [example 1](https://bitbucket.org/osrf/sdformat/issues/29/incompatibility-between-sdformats-urdf-and), [example 2](https://bugs.launchpad.net/ubuntu/+source/sdformat/+bug/1338845)) in different packages what can lead to really ugly bugs. You can check the list of people with broken systems in the [Robot Operative System issue](https://github.com/ros/rosdistro/issues/4633) and there is even one [guy asking for urdf packages for Brew back in 2014](https://github.com/ros/rosdistro/issues/4633#issuecomment-47642670). Let's not repeat old mistakes, please. Thanks!

File: Formula/dartsim.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "assimp"
+  depends_on "boost"
+  depends_on "eigen"
+  depends_on "fcl"
+  depends_on "libccd"
+  depends_on "nlopt"
+  depends_on "bullet"
```
Comment (fxcoudert) [inline Formula/dartsim.rb]: alphabetical ordering of runtime deps please

File: Formula/dartsim.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  test do
+    (testpath/"test.cpp").write <<~EOS
+      #include <dart/dart.hpp>
+      int main() {
+        auto world = std::make_shared<dart::simulation::World>();
+        assert(world != nullptr);
+        return 0;
+      }
+    EOS
+    system ENV.cc, "test.cpp", "-I#{include}/eigen3", "-L#{lib}", "-ldart", "-lassimp", "-lc++", "-std=c++11", "-o", "test"
```
Comment (fxcoudert) [inline Formula/dartsim.rb]: Why the `eigen3` inside `dartsim`'s own includes?

Comment (jslee02) [reply Formula/dartsim.rb]: `dartsim` depends on `eigen3`, and I believed this is a right way to set the include path of dependencies like [this](https://github.com/Homebrew/homebrew-core/blob/98768651244eeb1d63a3f48a2d1bda243dd86446/Formula/cgal.rb#L79).

Let me know if there is a recommended way.

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/dartsim.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  url "https://github.com/dartsim/dart/archive/v6.3.0.tar.gz"
+  sha256 "aa92634c1c97d99966cf16c4a0845792941358c063409fa00c28b4039c961c25"
+
+  depends_on "cmake" => :build
+  depends_on "pkg-config" => :build
+  depends_on "assimp"
+  depends_on "boost"
+  depends_on "bullet"
+  depends_on "eigen"
+  depends_on "fcl"
```
[the remark this answers is not in the document]
Comment (jslee02) [reply Formula/dartsim.rb]: There is a PR (https://github.com/Homebrew/homebrew-core/pull/22726) ! :smile:

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by jslee02** [comment]
@fxcoudert I was about to ask if `-I#{Formula["eigen"].include}` is the right approach, but you beat me! Thanks for the fix. :+1:

File: Formula/dartsim.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (fxcoudert) [reply Formula/dartsim.rb]: I've merged it

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @jslee02 for the pull requests!

**Pull request closed without merging**